### PR TITLE
Merge submission v2.0 into android-v2

### PR DESCRIPTION
**Pull request by anhappdev**
Part of #278

### CODE REVIEW

**Comment by anhappdev** [comment]
Here is the APK built using this branch:
[2022-04-13_PR287_mlperf_app.apk](https://drive.google.com/file/d/1LPNDnskz1UGUKvuH7vHDaE3h_MquMxbA/view?usp=sharing)

**Comment by anhappdev** [comment]
@d-uzlov Do you know why do we still have `Build and test iOS Flutter app` test for the `android-v2` branch?

**Comment by d-uzlov** [comment]
@anhappdev
We didn't have any limitations in the iOS build job on our Jenkins server.
I have fixed it, from this point on it should only trigger in PR to the master branch

**Comment by anhappdev** [comment]
@d-uzlov Great. Thanks.

**Comment by anhappdev** [comment]
Here is the APK built using this branch:
[2022-04-14_PR287_mlperf_app.apk](https://drive.google.com/file/d/1x_yzSgcsz8US-aHGf27u2Af5SY5wZ0XI/view?usp=sharing)

**Comment by anhappdev** [comment]
Currently blocked by this issue:
https://github.com/mlcommons/mobile_app_open/issues/278#issuecomment-1097883953

**Comment by anhappdev** [comment]
@d-uzlov Danil, can you take a look at the test `Build-and-test-Android-App-PR (mobile-app-build-290400)` to see why it's failed?

File: mobile_back_tflite/cpp/backend_tflite/tflite_c.cc [hunk 1]
```diff
@@ -205,7 +205,7 @@ mlperf_backend_ptr_t mlperf_backend_create(
     //   if it is not specified in settings.
     // If we don't use batching, we will use shards_num=1,
     //   which is the default value.
-    backend_data->shards_num = 2;
+    backend_data->shards_num = 4;
```
Comment (anhappdev) [inline mobile_back_tflite/cpp/backend_tflite/tflite_c.cc:208]: The app crashed when running Image Classification (offline) on my test Android phone (TFLite backend).
It looks like an Out Of Memory issue, caused by this change. Setting the shards_num back to 2 resolves the crash.

**Comment by anhappdev** [comment]
@freedomtan @Mostelk @mohitmundhragithub @ptt2panda 

Please test this build on your specific device to make sure that everything still works as expected.
[2022-04-28_PR287_mlperf_app.apk](https://drive.google.com/file/d/1jLJ8OLEy5MXqT3mnZlddcFkRQEpbjWQ5/view?usp=sharing)

**Comment by freedomtan** [comment]
On a Xiaomi Redmi K50 Pro (a Dimensity 9000 phone), it works as expected except the Offline one (mostly because of `shards_num`)

**Comment by freedomtan** [comment]
|   backend | status   | tested by   |
|---|---|---|
| Google |   |   |
| MediaTek  |  looks good except Offline (mostly because of `shards_num`) | @freedomtan    |
| Qualcomm |   |   | 
| Samsung| | |

**Comment by kcwu** [comment]
I tested on a Google Pixel 6 and the build works as expected.

**Comment by ptt2panda** [comment]
@freedomtan, Kuang-Che (@kcwu) will be also working on MLPerf optimization for google, thanks.

**Comment by mohitmundhragithub** [comment]
> @freedomtan @Mostelk @mohitmundhragithub @ptt2panda
> 
> Please test this build on your specific device to make sure that everything still works as expected. [2022-04-28_PR287_mlperf_app.apk](https://drive.google.com/file/d/1jLJ8OLEy5MXqT3mnZlddcFkRQEpbjWQ5/view?usp=sharing)

I tested this apk on Qualcomm devices, and it seems to be working as expected.

**Review by mohitmundhragithub — APPROVED** [review]

**Comment by freedomtan** [comment]
@AhmedTElthakeb and @fatihcakirs please help check code and setting merged by @anhappdev are correct.

**Comment by freedomtan** [comment]
Slower performance may be caused by the latency setting.

File: android/cpp/mlperf_driver.cc [hunk 2]
```diff
@@ -102,7 +102,7 @@ void MlperfDriver::RunMLPerfTest(const std::string& mode, int min_query_count,
   } else {
     // Run MLPerf in SingleStream mode by default.
     mlperf_settings.scenario = ::mlperf::TestScenario::SingleStream;
-    mlperf_settings.single_stream_expected_latency_ns = 1000000;
+    mlperf_settings.single_stream_expected_latency_ns = 800000;
```
Comment (freedomtan) [inline android/cpp/mlperf_driver.cc:105]: @anhappdev please make 900,000 one for Samsung backend (unified app for Samsung to test).

Comment (anhappdev) [reply android/cpp/mlperf_driver.cc:105]: @AhmedTElthakeb @fatihcakirs 
This is the unified APK built with `mlperf_settings.single_stream_expected_latency_ns = 900000`
[2022-05-10_PR287_mlperf_app.apk](https://drive.google.com/file/d/1pHvJu0AIoKxZNtY2qDmqSwdZ6v0Tdq-S/view?usp=sharing)

**Comment by freedomtan** [comment]
FYR, on a Samsung S21 5G SM-G991B I borrowed from a colleague, it crashed. That is, I confirmed what @anhappdev reported for the master branch. Please check if this is expected behavior.

**Comment by anhappdev** [comment]
> @anhappdev could you please test this APK on your side? [mlperf_app_PR287_900k_051022.zip](https://github.com/mlcommons/mobile_app_open/files/8682217/mlperf_app_PR287_900k_051022.zip)

Yes. It worked fine on my test device `Samsung SM-M315F`

**Comment by freedomtan** [comment]
@anhappdev will create another apk, then hopefully we can close this

**Comment by anhappdev** [comment]
Here is the new APK:
[2022-05-17_PR287_fdc313f_mlperf_app.apk](https://drive.google.com/file/d/1BiWW__mlUERML0kZmtYaoi7X9L3PbyqH/view?usp=sharing)

**Comment by freedomtan** [comment]
the [2022-05-17_PR287_fdc313f_mlperf_app.apk](https://drive.google.com/file/d/1BiWW__mlUERML0kZmtYaoi7X9L3PbyqH/view?usp=sharing) provided by @anhappdev still get slightly lower numbers.

@anhappdev will provide a united app without Qualcomm backend to test if that's the reason.

**Comment by anhappdev** [comment]
This one without Qualcomm backend:
[2022-05-24_PR287_fdc313f_mlperf_app_smgt.apk](https://drive.google.com/file/d/1fe128hVyG9GdNW_wk6Pfy5rB1fY2wAV4/view?usp=sharing)

**Comment by anhappdev** [comment]
Samsung (@AhmedTElthakeb) confirmed that the app now works as expected with this build
[2022-05-27_PR287_59e1501b_mlperf_app_qsmgt.apk](https://drive.google.com/file/d/1dVJAXy6YkgNFPsaPimvCv3KTQrUyW5lK/view)

The issue was out-dated binary lib files in the build process. So the current code is now safe to merge without changes.

**Review by freedomtan — APPROVED** [review]
LGTM

**Pull request merged**